certdb/certdb.go

```diff
@@ -0,0 +1,227 @@
+	VALUES ($1, $2, $3, $4, $5, $6, $7);`
+
+	selectSQL = `
+SELECT %s FROM certificates
+	WHERE (serial = $1);`
+
+	selectAllSQL = `
+SELECT %s FROM certificates;`
+
+	selectAllUnexpiredSQL = `
+SELECT %s FROM certificates
+WHERE CURRENT_TIMESTAMP < expiry;`
+
+	updateRevokeSQL = `
+UPDATE certificates
+	SET status='revoked', revoked_at=CURRENT_TIMESTAMP, reason=$1
+	WHERE (serial = $2);`
+
+	insertOCSPSQL = `
+INSERT INTO ocsp_responses (serial, body, expiry)
+    VALUES ($1, $2, $3);`
+
+	updateOCSPSQL = `
+UPDATE ocsp_responses
+    SET expiry=$3, body=$2
+	WHERE (serial = $1);`
+
+	selectAllUnexpiredOCSPSQL = `
+SELECT %s FROM ocsp_responses
+WHERE CURRENT_TIMESTAMP < expiry;`
+
+	selectOCSPSQL = `
+SELECT %s FROM ocsp_responses
+    WHERE (serial = $1);`
+)
+
+func wrapCertStoreError(err error) error {
+	if err != nil {
+		return cferr.Wrap(cferr.CertStoreError, cferr.Unknown, err)
+	}
+	return nil
+}
+
+// InsertCertificate puts a CertificateRecord into db.
+func InsertCertificate(db *sql.DB, cr *CertificateRecord) error {
+	res, err := db.Exec(
+		insertSQL,
+		cr.Serial,
+		cr.CALabel,
+		cr.Status,
+		cr.Reason,
+		cr.Expiry.UTC(),
+		cr.RevokedAt,
+		cr.PEM,
+	)
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := res.RowsAffected()
+
+	if numRowsAffected != 1 {
+		return wrapCertStoreError(fmt.Errorf("%d rows are affected, should be 1 row", numRowsAffected))
+	}
+
+	return nil
+}
+
+// GetCertificate gets a CertificateRecord indexed by serial.
+func GetCertificate(db *sql.DB, serial string) (*CertificateRecord, error) {
+	cr := new(CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectSQL, sqlstruct.Columns(*cr)), serial)
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		return cr, wrapCertStoreError(sqlstruct.Scan(cr, rows))
+	}
+	return nil, nil
+}
+
+// GetUnexpiredCertificates gets all unexpired certificate from db.
+func GetUnexpiredCertificates(db *sql.DB) (crs []*CertificateRecord, err error) {
+	cr := new(CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectAllUnexpiredSQL, sqlstruct.Columns(*cr)))
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		err = sqlstruct.Scan(cr, rows)
+		if err != nil {
+			return nil, wrapCertStoreError(err)
+		}
+		crs = append(crs, cr)
+	}
+
+	return crs, nil
+}
+
+// RevokeCertificate updates a certificate with a given serial number and marks it revoked.
+func RevokeCertificate(db *sql.DB, serial string, reasonCode int) error {
+	result, err := db.Exec(updateRevokeSQL, reasonCode, serial)
+
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := result.RowsAffected()
+
+	if numRowsAffected != 1 {
+		return wrapCertStoreError(fmt.Errorf("%d rows are affected, should be 1 row", numRowsAffected))
+	}
+
+	return nil
+}
+
+// InsertOCSP puts a new OCSPRecord into the db.
+func InsertOCSP(db *sql.DB, rr *OCSPRecord) error {
```

Comment:
for now, I would provide "upsert" without using the real sql upsert. 1. a quick peek into sqlite and postgresql docs gets me two different syntaxs again (not surprisingly).
I assume this is accurate and up to date. https://wiki.postgresql.org/wiki/UPSERT  
2. we don't need a strict simultaneous consistency between OCSP and certificate status. It's ok that a OCSP response still shows 'good' while the corresponding certificate is being revoked seconds ago, as long as the OCSP response catch up to be eventually consistent (within hours to days). So write race condition between CA and ocspgen on OCSP table is OK since we don't have write race condition on Certificate table and ocspgen will run periodically to let OCSP catch up .